File: server/chat_app_server/app/models/message.py
#!/usr/bin/env python3
"""
消息相关的数据模型
"""
from pydantic import BaseModel
from typing import Optional, List, Dict, Any
from datetime import datetime
from .database_factory import get_database
import asyncio
import uuid
import json

def row_to_dict(row) -> Dict[str, Any]:
    """将数据库行转换为字典"""
    if row is None:
        return None
    # SQLite Row对象可以直接转换为字典
    return {key: row[key] for key in row.keys()}

class MessageCreate(BaseModel):
    # 前端发送的字段名（驼峰命名）
    id: Optional[str] = None
    sessionId: str
    role: str
    content: str
    summary: Optional[str] = None
    toolCalls: Optional[List[Dict[str, Any]]] = None
    tool_call_id: Optional[str] = None
    # 用于持久化的推理字段（内部统一使用该字段）
    reasoning: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None
    createdAt: Optional[str] = None
    status: Optional[str] = None
    
    # 内部属性，用于数据库操作
    @property
    def session_id(self) -> str:
        return self.sessionId
    
    @session_id.setter
    def session_id(self, value: str):
        self.sessionId = value
    
    @property
    def tool_calls_str(self) -> Optional[str]:
        """将toolCalls转换为JSON字符串"""
        if self.toolCalls:
            return json.dumps(self.toolCalls)
        return None
    
    @property
    def metadata_str(self) -> Optional[str]:
        """将metadata转换为JSON字符串"""
        if self.metadata:
            return json.dumps(self.metadata)
        return None

    # 已不再需要统一属性，直接使用 reasoning 字段

    @classmethod
    async def create(cls, message_data: "MessageCreate") -> Dict[str, Any]:
        """创建新消息"""
        # 使用前端提供的ID，如果没有则生成新的
        message_id = message_data.id or str(uuid.uuid4())
        
        query = """
        INSERT INTO messages (id, session_id, role, content, summary, tool_calls, tool_call_id, reasoning, metadata, created_at)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, CURRENT_TIMESTAMP)
        """
        
        db = get_database()
        await db.execute_query_async(query, (
            message_id,
            message_data.session_id,  # 使用属性访问器
            message_data.role,
            message_data.content,
            message_data.summary,
            message_data.tool_calls_str,  # 使用JSON字符串转换
            message_data.tool_call_id,
            message_data.reasoning,
            message_data.metadata_str  # 使用JSON字符串转换
        ))
        
        return await cls.get_by_id(message_id)

    @classmethod
    def create_sync(cls, message_data: "MessageCreate") -> Dict[str, Any]:
        """创建新消息（同步版本）"""
        # 使用前端提供的ID，如果没有则生成新的
        message_id = message_data.id or str(uuid.uuid4())
        
        query = """
        INSERT INTO messages (id, session_id, role, content, summary, tool_calls, tool_call_id, reasoning, metadata, created_at)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, CURRENT_TIMESTAMP)
        """
        
        db = get_database()
        db.execute_sync(query, (
            message_id,
            message_data.session_id,  # 使用属性访问器
            message_data.role,
            message_data.content,
            message_data.summary,
            message_data.tool_calls_str,  # 使用JSON字符串转换
            message_data.tool_call_id,
            message_data.reasoning,
            message_data.metadata_str  # 使用JSON字符串转换
        ))
        
        return cls.get_by_id_sync(message_id)

    @classmethod
    def get_by_id_sync(cls, message_id: str) -> Optional[Dict[str, Any]]:
        """根据ID获取消息（同步版本）"""
        query = "SELECT * FROM messages WHERE id = ?"
        db = get_database()
        row = db.fetchone_sync(query, (message_id,))
        return row_to_dict(row)

    @classmethod
    async def get_by_session(cls, session_id: str, limit: Optional[int] = None) -> List[Dict[str, Any]]:
        """获取会话的所有消息"""
        query = "SELECT * FROM messages WHERE session_id = ? ORDER BY created_at ASC"
        params = (session_id,)
        
        if limit:
            query += " LIMIT ?"
            params = (session_id, limit)
        
        db = get_database()
        rows = await db.fetch_all_async(query, params)
        return [row_to_dict(row) for row in rows]

    @classmethod
    def get_by_session_sync(cls, session_id: str, limit: Optional[int] = None) -> List[Dict[str, Any]]:
        """获取会话的所有消息（同步版本）"""
        query = "SELECT * FROM messages WHERE session_id = ? ORDER BY created_at ASC"
        params = (session_id,)
        
        if limit:
            query += " LIMIT ?"
            params = (session_id, limit)
        
        db = get_database()
        rows = db.fetchall_sync(query, params)
        return [row_to_dict(row) for row in rows]

    @classmethod
    async def get_by_id(cls, message_id: str) -> Optional[Dict[str, Any]]:
        """根据ID获取消息"""
        query = "SELECT * FROM messages WHERE id = ?"
        db = get_database()
        row = await db.fetch_one_async(query, (message_id,))
        return row_to_dict(row)

    @classmethod
    async def delete_by_session(cls, session_id: str) -> bool:
        """删除会话的所有消息"""
        query = "DELETE FROM messages WHERE session_id = ?"
        db = get_database()
        cursor = await db.execute_query_async(query, (session_id,))
        return cursor.rowcount > 0

    @classmethod
    def delete_by_session_sync(cls, session_id: str) -> bool:
        """删除会话的所有消息（同步版本）"""
        query = "DELETE FROM messages WHERE session_id = ?"
        db = get_database()
        cursor = db.execute_sync(query, (session_id,))
        return cursor.rowcount > 0

class ChatStopRequest(BaseModel):
    session_id: str

File: server/chat_app_node_server/src/models/message.js
/**
 * Message 数据模型 - 消息管理
 * 复刻自 Python: app/models/message.py
 */

import { v4 as uuidv4 } from 'uuid';
import { getDatabaseSync } from './database-factory.js';

/**
 * Message 模型类
 */
export class Message {
  constructor(data) {
    this.id = data.id || uuidv4();
    this.sessionId = data.sessionId;
    this.role = data.role; // user, assistant, tool
    this.content = data.content;
    this.summary = data.summary || null;
    this.toolCalls = data.toolCalls || null;
    this.tool_call_id = data.tool_call_id || null;
    this.reasoning = data.reasoning || null;
    this.metadata = data.metadata || null;
    this.created_at = data.created_at || new Date().toISOString();
  }

  /**
   * 转换为数据库格式
   */
  toDb() {
    return {
      id: this.id,
      session_id: this.sessionId,
      role: this.role,
      content: this.content,
      summary: this.summary,
      tool_calls: this.toolCalls ? JSON.stringify(this.toolCalls) : null,
      tool_call_id: this.tool_call_id,
      reasoning: this.reasoning,
      metadata: this.metadata ? JSON.stringify(this.metadata) : null,
      created_at: this.created_at
    };
  }

  /**
   * 从数据库行创建实例
   */
  static fromDb(row) {
    if (!row) return null;
    return new Message({
      id: row.id,
      sessionId: row.session_id,
      role: row.role,
      content: row.content,
      summary: row.summary,
      toolCalls: row.tool_calls ? JSON.parse(row.tool_calls) : null,
      tool_call_id: row.tool_call_id,
      reasoning: row.reasoning,
      metadata: row.metadata ? JSON.parse(row.metadata) : null,
      created_at: row.created_at
    });
  }

  /**
   * 转换为 OpenAI API 格式
   */
  toOpenAI() {
    if (this.role === 'tool') {
      return {
        role: 'tool',
        tool_call_id: this.tool_call_id,
        content: this.content
      };
    }

    const message = {
      role: this.role,
      content: this.content
    };

    if (this.toolCalls && this.toolCalls.length > 0) {
      message.tool_calls = this.toolCalls;
    }

    return message;
  }
}

/**
 * Message 服务类
 */
export class MessageService {
  /**
   * 创建消息（异步）
   */
  static async create(messageData) {
    const message = new Message(messageData);
    const db = getDatabaseSync();
    const dbData = message.toDb();

    await db.execute(
      `INSERT INTO messages (id, session_id, role, content, summary, tool_calls, tool_call_id, reasoning, metadata, created_at)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      [
        dbData.id,
        dbData.session_id,
        dbData.role,
        dbData.content,
        dbData.summary,
        dbData.tool_calls,
        dbData.tool_call_id,
        dbData.reasoning,
        dbData.metadata,
        dbData.created_at
      ]
    );

    return message;
  }

  /**
   * 创建消息（同步）
   */
  static createSync(messageData) {
    const message = new Message(messageData);
    const db = getDatabaseSync();
    const dbData = message.toDb();

    db.executeSync(
      `INSERT INTO messages (id, session_id, role, content, summary, tool_calls, tool_call_id, reasoning, metadata, created_at)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      [
        dbData.id,
        dbData.session_id,
        dbData.role,
        dbData.content,
        dbData.summary,
        dbData.tool_calls,
        dbData.tool_call_id,
        dbData.reasoning,
        dbData.metadata,
        dbData.created_at
      ]
    );

    return message;
  }

  /**
   * 根据 ID 获取消息（异步）
   */
  static async getById(messageId) {
    const db = getDatabaseSync();
    const row = await db.fetchone('SELECT * FROM messages WHERE id = ?', [messageId]);
    return Message.fromDb(row);
  }

  /**
   * 获取会话的所有消息（异步）
   */
  static async getBySession(sessionId, limit = null, offset = 0) {
    const db = getDatabaseSync();
    let query = 'SELECT * FROM messages WHERE session_id = ? ORDER BY created_at ASC';
    const params = [sessionId];

    if (limit) {
      query += ' LIMIT ? OFFSET ?';
      params.push(limit, offset);
    }

    const rows = await db.fetchall(query, params);
    return rows.map(row => Message.fromDb(row));
  }

  /**
   * 获取会话的最近 N 条消息（异步）
   */
  static async getRecentBySession(sessionId, limit = 10) {
    const db = getDatabaseSync();
    const rows = await db.fetchall(
      `SELECT * FROM messages
       WHERE session_id = ?
       ORDER BY created_at DESC
       LIMIT ?`,
      [sessionId, limit]
    );
    return rows.reverse().map(row => Message.fromDb(row));
  }

  /**
   * 获取会话的最近 N 条消息（支持偏移量，用于分页“加载更多”）
   * 语义：按 created_at DESC 做分页，再反转为 ASC 返回
   */
  static async getRecentPageBySession(sessionId, limit = 10, offset = 0) {
    const db = getDatabaseSync();
    const rows = await db.fetchall(
      `SELECT * FROM messages
       WHERE session_id = ?
       ORDER BY created_at DESC
       LIMIT ? OFFSET ?`,
      [sessionId, limit, offset]
    );
    return rows.reverse().map(row => Message.fromDb(row));
  }

  /**
   * 删除消息（异步）
   */
  static async delete(messageId) {
    const db = getDatabaseSync();
    await db.execute('DELETE FROM messages WHERE id = ?', [messageId]);
  }

  /**
   * 删除会话的所有消息（异步）
   */
  static async deleteBySession(sessionId) {
    const db = getDatabaseSync();
    await db.execute('DELETE FROM messages WHERE session_id = ?', [sessionId]);
  }

  /**
   * 统计会话的消息数量（异步）
   */
  static async countBySession(sessionId) {
    const db = getDatabaseSync();
    const result = await db.fetchone(
      'SELECT COUNT(*) as count FROM messages WHERE session_id = ?',
      [sessionId]
    );
    return result.count;
  }

  /**
   * 将消息列表转换为 OpenAI API 格式
   */
  static toOpenAIFormat(messages) {
    return messages.map(msg => msg.toOpenAI());
  }
}
